File: Miscellaneous/Other/Review/Stage 6/Review 2/Part2/monotonic-stack/monotonic-stack.go
package main

import "fmt"

func nextGreaterElement(nums []int) []int {
	result := make([]int, len(nums))
	for key := range result {
		result[key] = -1
	}
	monotonicStack := make([]int, 0)

	for i := 0; i < len(nums); i++ {
		for len(monotonicStack) > 0 && nums[monotonicStack[len(monotonicStack)-1]] < nums[i] {
			top := monotonicStack[len(monotonicStack)-1]
			monotonicStack = monotonicStack[0 : len(monotonicStack)-1]
			result[top] = nums[i]
		}
		monotonicStack = append(monotonicStack, i)
	}

	return result
}

// test
func main() {
	nums := []int{2, 1, 2, 4, 3}
	fmt.Println(nextGreaterElement(nums)) // [4, 2, 4, -1, -1]
}

File: Data Structure & Algorithm/Review/Stage 9/Review 3/Part2/monotonic-stack/monotonic-stack.py
from typing import List


def nextGreaterElement(nums: List[int]) -> List[int]:
    monotonic_stack: List[int] = []
    result: List[int] = [-1] * len(nums)

    for i in range(0, len(nums)):
        while (
            len(monotonic_stack) > 0
            and nums[i] > nums[monotonic_stack[len(monotonic_stack) - 1]]
        ):
            top = monotonic_stack.pop()
            result[top] = nums[i]
        monotonic_stack.append(i)

    return result


# test
nums = [2, 1, 2, 4, 3]
print(nextGreaterElement(nums))
# [4, 2, 4, -1, -1]
